uPortal-rendering/src/main/java/org/apereo/portal/url/UrlSyntaxProviderImpl.java

```diff
@@ -1156,97 +1156,49 @@ public String generateUrl(HttpServletRequest request, IPortalUrlBuilder portalUr
-                url.addPath(PORTLET_PATH_PREFIX);
-                url.addPath(targetedPortletString);
-            }
-            // For normal render requests (generally multiple portlets on a page) add the targeted
-            // portlet as a parameter
-            else {
-                url.addParameter(PARAM_TARGET_PORTLET, targetedPortletString);
-            }
-
-            /*
-             * CSRF Prevention
-             *
-             * Add the Spring-managed CSRF token to requests that need them.  This list _should_
-             * include Action URLs only, but several Resource URLs are currently being used with
-             * POST requests in Apereo portlets.  We need to include Resource URLs as well, since
-             * (just now) we don't have the time to correct all those usages.  We should work to
-             * correct those cases, and remove handling of Resource URLs when we can.
-             */
-            if (UrlType.ACTION.equals(urlType) || UrlType.RESOURCE.equals(urlType)) {
-                final CsrfToken token = (CsrfToken) request.getAttribute(CSRF_PARAMETER_NAME);
-                if (token != null) {
-                    url.setParameter(token.getParameterName(), token.getToken());
-                }
-            }
+            this.populateUrlFields(
+                    url, urlType, urlState, request, urlNodeSyntaxHelper, targetedPortletWindowId);
 
         } else {
-            final String targetFolderId = portalUrlBuilder.getTargetFolderId();
-            final List<String> folderNames =
-                    urlNodeSyntaxHelper.getFolderNamesForLayoutNode(request, targetFolderId);
-            if (folderNames != null && !folderNames.isEmpty()) {
-                url.addPath(FOLDER_PATH_PREFIX);
-                for (final String folderName : folderNames) {
-                    url.addPath(folderName);
-                }
-            }
-
+            this.addFolderInformationForTargetFolder(
+                    request, portalUrlBuilder, urlNodeSyntaxHelper, url);
             urlState = UrlState.NORMAL;
             resourceId = null;
         }
 
-        // Add the state of the URL
-        url.addPath(urlState.toLowercaseString());
+        this.addUrlStateAndTypeToUrl(url, urlType, urlState, resourceId);
+        this.addPortalParametersToUrl(portalUrlBuilder, url);
+        this.addPortletsParametersForEachUrl(
+                request, url, portletUrlBuilders, targetedPortletWindowId, urlType, urlState);
 
-        // File part specifying the type of URL, resource URLs include the resourceId
-        if (urlType == UrlType.RESOURCE && resourceId != null) {
-            url.addPath(resourceId + "." + urlType.toLowercaseString() + REQUEST_TYPE_SUFFIX);
-        } else {
-            url.addPath(urlType.toLowercaseString() + REQUEST_TYPE_SUFFIX);
+        if (logger.isDebugEnabled()) {
+            logger.debug("Generated '" + url + "' from '" + portalUrlBuilder);
         }
```

Comment:
I'm not sure I follow this extraction, it is used only once.
Why pull it out?

Reply:
I believe that the extract method can be applied in case 1> the method is too long and / or the 2> the method is assigned multiple responsibilities.  (violating SRP principle as well)
In this case, I felt both these aspects are applicable (to a certain compelling extent).
Hence, I went ahead and applied extract method refactor.  This Refactoring results in smaller and more cohesive
method.